homeassistant/components/smart_meter_texas/__init__.py

```diff
@@ -0,0 +1,100 @@
+"""The Smart Meter Texas integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+from smart_meter_texas.async_api import Auth, Meter
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Required("meter"): str,
+                vol.Required("esiid"): str,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Smart Meter Texas component."""
+
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+
+    if not conf:
+        return True
+
+    hass.async_create_task(hass.config_entries.flow.async_init(DOMAIN))
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Smart Meter Texas from a config entry."""
+
+    entry_data = entry.data
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    async def async_update_data():
+        auth = Auth(websession, entry_data[CONF_USERNAME], entry_data[CONF_PASSWORD])
+        await auth.authenticate()
```

Comment:
We should avoid re-login if possible as these tend to be very expensive operations on an api which can lead to users being banned.

If you can't tell when the token / session / cookie expires one strategy is to relogin when you get an HTTP 401 from any other api call (only one time though as we don't want to pound the login end point with requests if their password changes)

Reply:
I believe the login token is only good for around 30 minutes. The meter read API only allows for 24 calls per day so I'm pretty sure it will require re-authenticating every request anyways. I'll take another look at it though and try like you suggested. It's an undocumented API so it's all trial and error.